File: app/src/main/java/com/shakil/barivara/utils/FilterManager.java
package com.shakil.barivara.utils;

import android.content.Context;

import com.shakil.barivara.model.meter.ElectricityBill;
import com.shakil.barivara.model.meter.Meter;
import com.shakil.barivara.model.room.Rent;
import com.shakil.barivara.model.room.Room;
import com.shakil.barivara.model.tenant.Tenant;

import java.util.ArrayList;

public class FilterManager{
    private Context context;

    public FilterManager(Context context) {
        this.context = context;
    }

    //region room search listener
    public interface onFilterClick{
        void onClick(ArrayList<Room> objects);
    }
    public void onFilterClick(String searchText, ArrayList<Room> sourceList, onFilterClick onFilterClick){
        ArrayList<Room> roomList = new ArrayList<>();
        for (int start = 0; start < sourceList.size(); start++) {
            if (sourceList.get(start).getRoomName().contains(searchText)){
                roomList.add(sourceList.get(start));
            }
        }
        if (onFilterClick != null){
            onFilterClick.onClick(roomList);
        }
    }
    //endregion

    //region rent search listener
    public interface onRentFilterClick{
        void onClick(ArrayList<Rent> objects);
    }
    public void onFilterClick(String searchText, ArrayList<Rent> sourceList, onRentFilterClick onRentFilterClick){
        ArrayList<Rent> rentList = new ArrayList<>();
        for (int start = 0; start < sourceList.size(); start++) {
            if (sourceList.get(start).getMonthName().contains(searchText)){
                rentList.add(sourceList.get(start));
            }
        }
        if (onRentFilterClick != null){
            onRentFilterClick.onClick(rentList);
        }
    }
    //endregion

    //region meter search listener
    public interface onMeterFilterClick{
        void onClick(ArrayList<Meter> objects);
    }
    public void onFilterClick(String searchText, ArrayList<Meter> sourceList, onMeterFilterClick onMeterFilterClick){
        ArrayList<Meter> meterList = new ArrayList<>();
        for (int start = 0; start < sourceList.size(); start++) {
            if (sourceList.get(start).getMeterName().contains(searchText)){
                meterList.add(sourceList.get(start));
            }
        }
        if (onMeterFilterClick != null){
            onMeterFilterClick.onClick(meterList);
        }
    }
    //endregion

    //region electricity bill search listener
    public interface onBillFilterClick{
        void onClick(ArrayList<ElectricityBill> objects);
    }
    public void onFilterClick(String searchText, ArrayList<ElectricityBill> sourceList, onBillFilterClick onBillFilterClick){
        ArrayList<ElectricityBill> billList = new ArrayList<>();
        for (int start = 0; start < sourceList.size(); start++) {
            if (sourceList.get(start).getRoomName().contains(searchText)){
                billList.add(sourceList.get(start));
            }
        }
        if (onBillFilterClick != null){
            onBillFilterClick.onClick(billList);
        }
    }
    //endregion

    //region tenant search listener
    public interface onTenantFilterClick{
        void onClick(ArrayList<Tenant> objects);
    }
    public void onFilterClick(String searchText, ArrayList<Tenant> sourceList, onTenantFilterClick onTenantFilterClick){
        ArrayList<Tenant> billList = new ArrayList<>();
        for (int start = 0; start < sourceList.size(); start++) {
            if (sourceList.get(start).getTenantName().contains(searchText)){
                billList.add(sourceList.get(start));
            }
        }
        if (onTenantFilterClick != null){
            onTenantFilterClick.onClick(billList);
        }
    }
    //endregion
}

File: app/src/main/java/com/shakil/barivara/utils/FilterManager.kt
package com.shakil.barivara.utils

import com.shakil.barivara.data.model.meter.ElectricityBill
import com.shakil.barivara.data.model.meter.Meter
import com.shakil.barivara.data.model.room.NewRoom
import com.shakil.barivara.data.model.room.Rent
import com.shakil.barivara.data.model.tenant.Tenant
import java.util.Locale

class FilterManager {
    fun onFilterClick(
        searchText: String,
        sourceList: List<NewRoom>,
        onFilterClick: onFilterClick
    ) {
        val roomList = ArrayList<NewRoom>()
        for (start in sourceList.indices) {
            if (sourceList[start].name.orEmpty().lowercase(Locale.getDefault()).contains(
                    searchText.lowercase(
                        Locale.getDefault()
                    )
                )
            ) {
                roomList.add(sourceList[start])
            }
        }
        onFilterClick.onClick(roomList)
    }

    fun onFilterClick(
        searchText: String,
        sourceList: ArrayList<Rent>,
        onRentFilterClick: onRentFilterClick
    ) {
        val rentList = ArrayList<Rent>()
        for (start in sourceList.indices) {
            if (sourceList[start].monthName.lowercase(Locale.getDefault()).contains(
                    searchText.lowercase(
                        Locale.getDefault()
                    )
                )
            ) {
                rentList.add(sourceList[start])
            }
        }
        onRentFilterClick.onClick(rentList)
    }

    fun onFilterClick(
        searchText: String,
        sourceList: ArrayList<Meter>,
        onMeterFilterClick: onMeterFilterClick
    ) {
        val meterList = ArrayList<Meter>()
        for (start in sourceList.indices) {
            if (sourceList[start].meterName.lowercase(Locale.getDefault()).contains(
                    searchText.lowercase(
                        Locale.getDefault()
                    )
                )
            ) {
                meterList.add(sourceList[start])
            }
        }
        onMeterFilterClick.onClick(meterList)
    }

    fun onFilterClick(
        searchText: String,
        sourceList: ArrayList<ElectricityBill>,
        onBillFilterClick: onBillFilterClick
    ) {
        val billList = ArrayList<ElectricityBill>()
        for (start in sourceList.indices) {
            if (sourceList[start].roomName.lowercase(Locale.getDefault()).contains(
                    searchText.lowercase(
                        Locale.getDefault()
                    )
                )
            ) {
                billList.add(sourceList[start])
            }
        }
        onBillFilterClick.onClick(billList)
    }

    fun onFilterClick(
        searchText: String,
        sourceList: List<Tenant>,
        onTenantFilterClick: onTenantFilterClick?
    ) {
        val billList = ArrayList<Tenant>()
        for (start in sourceList.indices) {
            if (sourceList[start].name.lowercase(Locale.getDefault()).contains(
                    searchText.lowercase(
                        Locale.getDefault()
                    )
                )
            ) {
                billList.add(sourceList[start])
            }
        }
        onTenantFilterClick?.onClick(billList)
    }

    interface onRentFilterClick {
        fun onClick(objects: ArrayList<Rent>)
    }

    interface onFilterClick {
        fun onClick(objects: ArrayList<NewRoom>)
    }

    interface onMeterFilterClick {
        fun onClick(objects: ArrayList<Meter>)
    }

    interface onTenantFilterClick {
        fun onClick(objects: ArrayList<Tenant>)
    }

    interface onBillFilterClick {
        fun onClick(objects: ArrayList<ElectricityBill>)
    }
}
